pkg/vm/stackitem/json.go

```diff
@@ -260,72 +261,110 @@ func (d *decoder) decodeMap() (*Map, error) {
 
 // ToJSONWithTypes serializes any stackitem to JSON in a lossless way.
 func ToJSONWithTypes(item Item) ([]byte, error) {
-	result, err := toJSONWithTypes(item, make(map[Item]bool, typicalNumOfItems))
-	if err != nil {
-		return nil, err
-	}
-	return json.Marshal(result)
+	return toJSONWithTypes(nil, item, make(map[Item]sliceNoPointer, typicalNumOfItems))
 }
 
-func toJSONWithTypes(item Item, seen map[Item]bool) (interface{}, error) {
+func toJSONWithTypes(data []byte, item Item, seen map[Item]sliceNoPointer) ([]byte, error) {
+	if item == nil {
+		return nil, fmt.Errorf("%w: nil", ErrUnserializable)
+	}
 	if len(seen) > MaxJSONDepth {
-		return "", ErrTooDeep
+		return nil, ErrTooDeep
+	}
+	if old, ok := seen[item]; ok {
+		if old.end == 0 {
+			// Compound item marshaling which has not yet finished.
+			return nil, ErrRecursive
+		}
+		if len(data)+old.end-old.start > MaxSize {
+			return nil, errTooBigSize
+		}
+		return append(data, data[old.start:old.end]...), nil
+	}
+
+	if _, ok := item.(Null); ok {
+		if len(data)+14 > MaxSize {
+			return nil, errTooBigSize
+		}
+		data = append(data, `{"type":"Any"}`...)
+		return data, nil
+	}
+
+	start := len(data)
+
+	prefix := `{"type":"` + item.Type().String() + `","value":`
+	if len(data)+len(prefix) > MaxSize {
+		return nil, errTooBigSize
 	}
-	var value interface{}
+	data = append(data, prefix...)
+
+	// Primitive stack items are appended after the switch
+	// to reduce the amount of size checks.
+	var primitive string
+	var err error
+
 	switch it := item.(type) {
 	case *Array, *Struct:
-		if seen[item] {
-			return "", ErrRecursive
-		}
-		seen[item] = true
-		arr := []interface{}{}
-		for _, elem := range it.Value().([]Item) {
-			s, err := toJSONWithTypes(elem, seen)
+		seen[item] = sliceNoPointer{start, 0}
+		data = append(data, '[')
+		for i, elem := range it.Value().([]Item) {
+			if i != 0 {
+				data = append(data, ',')
+			}
+			data, err = toJSONWithTypes(data, elem, seen)
 			if err != nil {
-				return "", err
+				return nil, err
 			}
-			arr = append(arr, s)
 		}
-		value = arr
-		delete(seen, item)
 	case Bool:
-		value = bool(it)
+		if it {
+			primitive = "true"
+		} else {
+			primitive = "false"
+		}
 	case *Buffer, *ByteArray:
-		value = base64.StdEncoding.EncodeToString(it.Value().([]byte))
+		primitive = `"` + base64.StdEncoding.EncodeToString(it.Value().([]byte)) + `"`
```

Comment:
BTW, `*Buffer` is a reference type, so it might be cached in `seen` also.

Reply:
Fixed.

@AnnaShaleva it was used this way before. Now we use `seen[item].end == 0` as a marker. This doesn't lead to a problem with `Buffer`, because there will be no recursion and `end` is set to non-zero value before the function exits.

Reply:
Oh, sorry, I understand what you mean now.